### PR TITLE
QCQMC notebooks

### PR DESCRIPTION
**Pull request by fdmalone**
Adds 3 notebooks for the QCQMC paper:

1. Code overview
2. End-to-end example
3. Experimental data reproduction

Depends on #356

### CODE REVIEW

**Comment by fdmalone** [comment]
@wjhuggins PTAL and feel free to add any details you like (high or low level) we can also revisit to add more helpful tutorials for shadow tomography..

**Comment by fdmalone** [comment]
@mpharrigan do I need to do anything more to ensure the docs render correctly / regenerate them?

File: dev_tools/check_notebooks.py
Comment (mpharrigan) [inline dev_tools/check_notebooks.py:1]: why have all the quotes changed  to quotes. this clutters up the diff

Comment (fdmalone) [reply dev_tools/check_notebooks.py:1]: Whoops, sorry, aggressive auto-formatter (will remove)

**Comment by mpharrigan** [comment]
I think you need to list the new notebooks in `toc.yaml`

File: dev_tools/check_notebooks.py
Comment (fdmalone) [reply dev_tools/check_notebooks.py:1]: Fixed.

**Comment by fdmalone** [comment]
I can also split this up if it's more digestible

File: recirq/qcqmc/hamiltonian.py [hunk 1]
```diff
@@ -84,9 +84,6 @@ class HamiltonianFileParams(data.Params):
         n_orb: The number of spatial orbitals.
         n_elec: The total number of electrons.
         do_eri_restore: Whether to restore 8-fold symmetry to the integrals.
-        path_prefix: An optional path string to prepend to the default
```
Comment (mpharrigan) [inline recirq/qcqmc/hamiltonian.py:87]: this is still an attribute / arg. Can you keep & update the docstring?

File: recirq/qcqmc/hamiltonian.py [hunk 2]
```diff
@@ -229,6 +221,107 @@ def get_restricted_fqe_hamiltonian(self) -> RestrictedHamiltonian:
...
+            pyscf_molecule=molecule, pyscf_scf=pyscf_scf
+        )
+
+        e_core = float(molecule.energy_nuc())
+
+        pyscf_fci = fci.FCI(molecule, pyscf_scf.mo_coeff)
+        pyscf_fci.verbose = 0
+        e_fci = pyscf_fci.kernel()[0]
+
+        return HamiltonianData(
```
Comment (mpharrigan) [inline recirq/qcqmc/hamiltonian.py:316]: nit: you can return `cls(...)` to theoretically support inheritance

File: recirq/qcqmc/hamiltonian.py [hunk 3]
```diff
@@ -229,6 +221,107 @@ def get_restricted_fqe_hamiltonian(self) -> RestrictedHamiltonian:
...
+            params=params,
+            e_core=e_core,
+            one_body_integrals=one_body_integrals,
+            two_body_integrals_pqrs=two_body_integrals_pqrs,
+            e_hf=e_hf,
+            e_fci=e_fci,
+        )
+
+    @classmethod
+    def build_hamiltonian_from_pyscf(
```
Comment (mpharrigan) [inline recirq/qcqmc/hamiltonian.py:273]: nit: can probably simplify the method name since you know it will be prepended with `HamiltonianData.`

**Review by mpharrigan — APPROVED** [review]

**Comment by fdmalone** [comment]
In terms of notebooks should these be cleared of output. Is the website updated automatically?

File: docs/qcqmc/full_workflow.ipynb [hunk 4]
```diff
@@ -0,0 +1,363 @@
+{
+ "cells": [
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# End-to-End Example\n",
+    "\n",
+    "We learned in the [high-level](high-level.ipynb) notebook how to setup the quantum side of the calculation, which involves specifying a Hamiltonian, defining a trial wavefunction ansatz before performing shadow tomography to extract the AFQMC trial wavefunction. In this example we will repeat these steps again but now for the case of H4 which was studied in the [paper](https://arxiv.org/pdf/2106.16235.pdf). Then we will outline how to interface the output of the quantum half of the calculation with [ipie](https://github.com/JoonhoLee-Group/ipie/) to perform AFQMC with the quantum trial wavefunction."
```
Comment (dstrain115) [inline docs/qcqmc/full_workflow.ipynb:10]: nit: Comma after "In this example".  I would probably also remove "but now" to make the sentence simpler.

File: docs/qcqmc/full_workflow.ipynb [hunk 5]
```diff
@@ -0,0 +1,363 @@
...
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Build Perfect Pairing Wavefunction\n",
+    "\n",
+    "Next we build a perfect pairing Hamiltonian based upon our `pyscf_params`"
```
Comment (dstrain115) [inline docs/qcqmc/full_workflow.ipynb:53]: Is "perfect pairing Hamiltonian" defined elsewhere?  I personally don't know what that is.

File: docs/qcqmc/full_workflow.ipynb [hunk 6]
```diff
@@ -0,0 +1,363 @@
...
+    "print(f\"Exact ground state energy: {trial_wf.fci_energy}\")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from recirq.qcqmc.blueprint import BlueprintParamsTrialWf, BlueprintData\n",
```
Comment (dstrain115) [inline docs/qcqmc/full_workflow.ipynb:101]: Should there be a description for this cell?  There's a bunch of cells in a row here with no description besides the short one at the top of the section.

File: docs/qcqmc/full_workflow.ipynb [hunk 7]
```diff
@@ -0,0 +1,363 @@
...
+    "analysis = OverlapAnalysisData.build_analysis_from_dependencies(\n",
+    "    analysis_params, dependencies=all_dependencies\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Finally we save the wavefunction in a format that is acceptable for ipie."
```
Comment (dstrain115) [inline docs/qcqmc/full_workflow.ipynb:155]: nit: I think there should be a comma after "Finally"

File: docs/qcqmc/full_workflow.ipynb [hunk 8]
```diff
@@ -0,0 +1,363 @@
...
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Run AFQMC \n",
+    "\n",
+    "Now that we have built our quantum wavefunction we can use it as a trial wavefunction in an AFQMC simulation. First we need to build a factorized Hamiltonian which is required for AFQMC. \n",
```
Comment (dstrain115) [inline docs/qcqmc/full_workflow.ipynb:181]: Comma after wavefunction.  I think there should be one after First too.

File: docs/qcqmc/full_workflow.ipynb [hunk 9]
```diff
@@ -0,0 +1,363 @@
...
+    "\n",
+    "In particular we require the three-index Cholesky tensor (`ham.chol` below) which satisfies\n",
```
Comment (dstrain115) [inline docs/qcqmc/full_workflow.ipynb:183]: Comma after "In particular"

File: docs/qcqmc/full_workflow.ipynb [hunk 10]
```diff
@@ -0,0 +1,363 @@
...
+    "assert ham.H1.shape == (2, pyscf_params.n_orb, pyscf_params.n_orb)\n",
+    "assert ham.chol.shape[0] == pyscf_params.n_orb * pyscf_params.n_orb"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Next we need to build a trial wavefunction from quantum wavefunction. In practice the quantum trial is expanded as a linear combination of (orthogonal) Slater Determinants. Within ipie this style of trial wavefunction is defined as a `ParticleHole` trial wavefunction. "
```
Comment (dstrain115) [inline docs/qcqmc/full_workflow.ipynb:212]: Comma after "Next",  "In practice" , and also "Within ipie"

File: docs/qcqmc/full_workflow.ipynb [hunk 11]
```diff
@@ -0,0 +1,363 @@
...
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Note that the variational energy might differ slightly from the result compute from the previous section. This is because the cholesky factorization uses a threshold of $1\\times10^{-5}$ a stopping criteria for convergence. Reducing the parameter `chol_cut` in `generate_hamiltonian_from_chk` will yield better agreement.\n",
+    "\n",
+    "At this point we are ready to run AFQMC. We need to build an `AFQMC` driver class which takes as input the factorized Hamiltonian and our `ParticleHole` trial wavefunction."
```
Comment (dstrain115) [inline docs/qcqmc/full_workflow.ipynb:250]: Comma after "At this point"

File: docs/qcqmc/full_workflow.ipynb [hunk 12]
```diff
@@ -0,0 +1,363 @@
...
+    "    ipie_path.mkdir(parents=True)\n",
+    "qmc_driver.run(estimator_filename=f'{ipie_path}/estimates.h5')"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Next we can visualize the data which is by default saved to `estimates.0.h5`."
```
Comment (dstrain115) [inline docs/qcqmc/full_workflow.ipynb:283]: Comma after next.

File: docs/qcqmc/full_workflow.ipynb [hunk 13]
```diff
@@ -0,0 +1,363 @@
...
+    "plt.xlabel(\"Block number\")\n",
+    "plt.ylabel(\"Total Energy (Ha)\")"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Through visual inspection we can determine when the calculation has equilibrated, after which we should be sampling the (approximate) ground state. We can use samples from this point on to estimate the AFQMC energy and error bar. The function `reblock_minimal` will perform the necessary error analysis taking into account the serial temporal correlation in the AFQMC data. "
```
Comment (dstrain115) [inline docs/qcqmc/full_workflow.ipynb:307]: Comma after "Through visual inspection"

File: docs/qcqmc/full_workflow.ipynb [hunk 14]
```diff
@@ -0,0 +1,363 @@
...
+   "source": [
+    "from ipie.analysis.blocking import reblock_minimal \n",
+    "reblock_minimal(qmc_driver.estimators.filename, start_block=20)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Note that the number of walkers and the number of blocks is probably too low to obtain statistically significant results and we would advise increasing both of these. For larger scale simulations would can use MPI to distribute the work among many MPI processes. See [ipie](https://github.com/JoonhoLee-Group/ipie/) for further details."
```
Comment (dstrain115) [inline docs/qcqmc/full_workflow.ipynb:324]: Comma after "For larger scale simulations".  Also, the following word is not correct.  Did you mean "we can use MPI to ..."

File: docs/qcqmc/high_level.ipynb [hunk 15]
```diff
@@ -0,0 +1,286 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "id": "679c5fd1",
+   "metadata": {},
+   "source": [
+    "# QCQMC\n",
+    "\n",
+    "This notebooks demonstrates the high-level code interface used to run the quantum part of https://arxiv.org/pdf/2106.16235.pdf.\n",
```
Comment (dstrain115) [inline docs/qcqmc/high_level.ipynb:10]: "This notebook" or "These notebooks"

File: docs/qcqmc/high_level.ipynb [hunk 16]
```diff
@@ -0,0 +1,286 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "00d220ee",
+   "metadata": {},
+   "source": [
+    "## Hamiltonian\n",
+    "\n",
+    "The first step is to pick a molecule and generate or load its relevant physical properties, namely the Hamiltonian. Here we specify a 2-electron Fermi-Hubbard Hamiltonian in the sto3g basis. Integral data is stored in the data/ file in this repository.\n",
```
Comment (dstrain115) [inline docs/qcqmc/high_level.ipynb:37]: What is the sto3g basis?  Is this a common terminology?  Can we link to a wikipedia or add a short description to explain it?

File: docs/qcqmc/high_level.ipynb [hunk 17]
```diff
@@ -0,0 +1,286 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "74230882",
+   "metadata": {},
+   "source": [
+    "## Trial Wavefunction\n",
+    "\n",
+    "Next, we specify a trial wavefunction. Here: we request a perfect pairing trial (using the specialized Params class) and don't include any heuristic layers (to keep the example simple and the runtime short).\n",
```
Comment (dstrain115) [inline docs/qcqmc/high_level.ipynb:78]: Remove colon after "Here"

File: recirq/qcqmc/trial_wf.py [hunk 18]
```diff
@@ -212,6 +213,97 @@ def _json_dict_(self):
...
+        params: PerfectPairingPlusTrialWavefunctionParams,
+        *,
+        dependencies: Dict[data.Params, data.Data],
+        do_print: bool = False,
+    ) -> "TrialWavefunctionData":
+        """Builds a TrialWavefunctionData from a TrialWavefunctionParams
+
+        Args:
+            params: The parameters specifying the PP+ trial wavefunction.
+            dependencies: Data dependencies
```
Comment (dstrain115) [inline recirq/qcqmc/trial_wf.py:228]: Can we add more information to this arg?

File: recirq/qcqmc/trial_wf.py [hunk 19]
```diff
@@ -212,6 +213,97 @@ def _json_dict_(self):
...
+            do_print: Print debugging information to stdout
```
Comment (dstrain115) [inline recirq/qcqmc/trial_wf.py:229]: nit: Change to "If set to True, print debugging information to stdout"

File: recirq/qcqmc/trial_wf.py [hunk 20]
```diff
@@ -212,6 +213,97 @@ def _json_dict_(self):
...
+
+        Returns:
+            The constructed TrialWavefunctionData object.
+        """
+
+        # Avoid circular imports
+        from recirq.qcqmc import optimize_wf
```
Comment (dstrain115) [inline recirq/qcqmc/trial_wf.py:236]: This is a code smell that indicates that we may want to further split up these files in order to remove this.  You can do this in a separate PR if you want.

File: recirq/qcqmc/trial_wf.py [hunk 21]
```diff
@@ -212,6 +213,97 @@ def _json_dict_(self):
...
+                do_print=do_print,
+                use_fast_gradients=params.use_fast_gradients,
+            )
+        else:
+            if (
+                params.initial_two_body_qchem_amplitudes is None
+                or params.initial_orbital_rotation is not None
+            ):
+                raise NotImplementedError(
+                    "TODO: Implement whatever isn't finished here."
```
Comment (dstrain115) [inline recirq/qcqmc/trial_wf.py:269]: What?  Let's at least put a better message here.

File: recirq/qcqmc/hamiltonian.py (hunk 1 above)
Comment (fdmalone) [reply recirq/qcqmc/hamiltonian.py:87]: Done.

File: recirq/qcqmc/hamiltonian.py (hunk 2 above)
Comment (fdmalone) [reply recirq/qcqmc/hamiltonian.py:316]: Done.

File: recirq/qcqmc/hamiltonian.py (hunk 3 above)
Comment (fdmalone) [reply recirq/qcqmc/hamiltonian.py:273]: do you mean remove the "_from"?

File: docs/qcqmc/full_workflow.ipynb (hunk 5 above)
Comment (fdmalone) [reply docs/qcqmc/full_workflow.ipynb:53]: should be `perfect paring wavefunction`. Fixed.

File: docs/qcqmc/full_workflow.ipynb (hunk 4 above)
Comment (fdmalone) [reply docs/qcqmc/full_workflow.ipynb:10]: Done.

File: docs/qcqmc/full_workflow.ipynb (hunk 6 above)
Comment (fdmalone) [reply docs/qcqmc/full_workflow.ipynb:101]: Added some description.

File: docs/qcqmc/full_workflow.ipynb (hunk 7 above)
Comment (fdmalone) [reply docs/qcqmc/full_workflow.ipynb:155]: Done.

File: docs/qcqmc/full_workflow.ipynb (hunk 8 above)
Comment (fdmalone) [reply docs/qcqmc/full_workflow.ipynb:181]: Done.

File: docs/qcqmc/full_workflow.ipynb (hunk 9 above)
Comment (fdmalone) [reply docs/qcqmc/full_workflow.ipynb:183]: Done.

File: docs/qcqmc/full_workflow.ipynb (hunk 10 above)
Comment (fdmalone) [reply docs/qcqmc/full_workflow.ipynb:212]: Done, done, done.

File: docs/qcqmc/full_workflow.ipynb (hunk 11 above)
Comment (fdmalone) [reply docs/qcqmc/full_workflow.ipynb:250]: Done.

File: docs/qcqmc/full_workflow.ipynb (hunk 12 above)
Comment (fdmalone) [reply docs/qcqmc/full_workflow.ipynb:283]: Done.

File: docs/qcqmc/full_workflow.ipynb (hunk 13 above)
Comment (fdmalone) [reply docs/qcqmc/full_workflow.ipynb:307]: Done.

File: docs/qcqmc/full_workflow.ipynb (hunk 14 above)
Comment (fdmalone) [reply docs/qcqmc/full_workflow.ipynb:324]: Done.

File: docs/qcqmc/high_level.ipynb (hunk 15 above)
Comment (fdmalone) [reply docs/qcqmc/high_level.ipynb:10]: -> This notebook.

File: docs/qcqmc/high_level.ipynb (hunk 16 above)
Comment (fdmalone) [reply docs/qcqmc/high_level.ipynb:37]: It's common for chemists. I added a link to wikipedia.

File: docs/qcqmc/high_level.ipynb (hunk 17 above)
Comment (fdmalone) [reply docs/qcqmc/high_level.ipynb:78]: Done.

File: recirq/qcqmc/trial_wf.py (hunk 18 above)
Comment (fdmalone) [reply recirq/qcqmc/trial_wf.py:228]: Done.

File: recirq/qcqmc/trial_wf.py (hunk 19 above)
Comment (fdmalone) [reply recirq/qcqmc/trial_wf.py:229]: Done.

File: recirq/qcqmc/trial_wf.py (hunk 20 above)
Comment (fdmalone) [reply recirq/qcqmc/trial_wf.py:236]: Agreed. I'll punt to a follow up.

File: recirq/qcqmc/trial_wf.py (hunk 21 above)
Comment (fdmalone) [reply recirq/qcqmc/trial_wf.py:269]: I also got confused by the logic reading it again. I think I simplified the error message.

**Comment by mpharrigan** [comment]
Yes, the website should execute the notebooks

**Comment by fdmalone** [comment]
@dstrain115 is this approved?

File: docs/qcqmc/experimental_wavefunctions.ipynb [hunk 22]
```diff
@@ -0,0 +1,203 @@
...
+    "| 387 | -1.947362 | -1.934682 | -1.910477 | -1.901883 |\n",
+    "| 652 | -1.952416 | -1.939853 | -1.912790 | -1.905250 |\n",
+    "| 1100 | -1.955544 | -1.944651 | -1.915073 | -1.909122 |\n",
+    "| 1856 | -1.955028 | -1.945966 | -1.909558 | -1.908038 |\n",
+    "| 3129 | -1.953877 | -1.947763 | -1.913386 | -1.908835 |\n",
+    "| 5276 | -1.954697 | -1.947323 | -1.912284 | -1.909315 |\n",
+    "| 8896 | -1.954930 | -1.947458 | -1.913889 | -1.913068 |\n",
+    "| 15000 | -1.954356 | -1.948894 | -1.913894 | -1.913082 |\n"
+   ]
+  },
```
Comment (dstrain115) [inline docs/qcqmc/experimental_wavefunctions.ipynb:68]: This notebook still has several code blocks in a row with almost no explanatory text.  It would be good if we can expound on what's going on a little bit more.

**Review by dstrain115 — APPROVED** [review]
One of the notebooks could use some more explanation.  After that, LGTM.
